### ISSUE TITLE
Should the "virtual environment" for third-party packages install all dependencies each time?

### ISSUE THREAD

**Issue by bmcdonald3** [chapel-lang/chapel#18370]
## Summary
Currently, running `make test-venv` will install all 3 of the `test-venv`, `c2chapel`, and `chpldoc` requirements even when they aren't intended to be used. Due to a recent package added to `c2chapel` (pycparserext), this requires that the user have Python 3.7. It seems reasonable that users should only need to download the packages they intend to use, so splitting them up into 3 separate installs may make sense.

## Positions
To summarize a few paths forward (I apologize if I poorly represent your position):
### Position 1: Do nothing
Since the virtual environment in question isn't something a typical user would use and is pretty much only used by developers, this does not seem worth doing and it is reasonable to expect a developer to have/install Python 3.7. 

#### Pros:
- No work for something with a (seemingly?) low payoff
- No risk of breaking anything

#### Cons:
- Requires unnecessary package downloads
- Forces all users of `test-venv` to have Python 3.7

### Position 2: Only download the required virtual environment packages
Since someone who runs `make test-venv`, for example, is not likely to run `c2chapel` it would make sense to separate these out and not install a bunch of packages that haven't been indicated would be used. The way that we install these packages is not literally in a "virtual environment", but is actually just a single Python file with all of the installs, so splitting them out is not as simple as just changing the `Makefile`

#### Pros:
- Allows people to only download what they need and not randomly install `sphinx` (which can take a while)

#### Cons:
- Would need to rethink the way we handle the installs (which could be easy or difficult, I don't think it is clear yet)

### Background
While running on a particular machine that didn't have Python 3.7 in it's path by default, a `make test-venv` was failing due to not having Python 3.7, even though the person installing `test-venv` had no intention of running c2chapel; in other words, Python 3.7 was required even though it wasn't at all related to what needed to be done. (See: https://github.com/Cray/chapel-private/issues/2480#issuecomment-912715266)

**Comment by bradcray** [comment]
Position 2 matches my intuition (where I'm motivated more by "install no more than is required" minimalism than the specifics of which Python versions are required); but I'm admittedly very ignorant of Python's virtual environments and the like.

I thought we actually used to have something more like position 2, making me wonder whether something required us to abandon it?  (though maybe I'm just misremembering).

**Comment by lydia-duncan** [comment]
Note that we have installed the requirements for these packages "separately" (but in the same virtual environment) in the past, but if my memory serves in switching to Python 3 we significantly redid our installation process, resulting in what we have today.

**Comment by mppf** [comment]
Right. We don't use a virtualenv/venv anymore and instead create a Python application. See also PR #16694. I think we could, in principle, have a few different Python applications, but IIRC that seemed too annoying to me to be worthwhile when I was working on that PR (After suffering through quite a lot of Python 3 issues). I don't think we should have a single Python application that has different contents in different configurations.

While in principle I can agree with not trying to get unused dependencies, in this specific case I appreciate the simplicity of a single Python application containing all of our dependencies. I also would prefer to have all of chpldoc/c2chapel/start_test requiring the same Python versions, FWIW.

As an aside, https://chapel-lang.org/docs/main/usingchapel/prereqs.html#readme-prereqs says we need Python 3.6 or newer but my understanding is that after the pycparserext change now it is 3.7 that is required.

**Comment by bradcray** [comment]
Thanks for the additional detail about the evolution of the code base here.

So maybe a related question then:  Is requiring Python 3.7 for any of these capabilities asking too much of our users?  Would we want to pull the `pycparserext` capability from `c2chapel` (which, for context, permits `c2chapel` to handle GNU extensions in header files) to maintain the previous prerequisites?

**Comment by mppf** [comment]
Python 3.6 is end-of-life'd in two months and so I think it's OK to require 3.7. Additionally I think it's OK to require a relatively recent Python for these add-on tools (c2chapel/chpldoc/start_test) that we don't require for basic `chpl` use.

**Comment by bradcray** [comment]
OK, so based on that, I'm fine with position 1 for now, though position 2 definitely sounds more attractive in the long-term.  But it sounds like we'll need to find someone who feels inspired to tackle this, and I'd be reluctant to have anyone try in the 1.25 timeframe given how painful Python changes have been in our testing in the past (unless it's less difficult / impactful than it sounds).

It does sound like we should update the prereqs file, for now, in any case, though, right?  @bmcdonald3 can you handle that?

**Comment by lydia-duncan** [comment]
If we do decide we want to keep position 1 for all time, I think it would make sense to rename the build target to reflect that it builds for more than just the testing environment.
